Add RV3032 MFD drive to expose Timer counter and Temperature sensor

bogdanovs opened the pull request:
This PR introduce MFD driver for RV3032 to expose Temperature sensor and Timer counter.

- MFD driver is taking gave for communication with the device via I2C and handling IRQs which are dispached accordingly
- Temperature sensor is implementing sensor API
- Timer counter is implementing couter API
- Existing RTC driver is reworked to match this

RV3032 RTC driver changes:
- IRQ is not handled by the driver anymore but MFD take ownership over that
- I2C call are made via `mfd_rv3032`  prefixed functions from MFD guarded from semaphor
- There was unclear behavior of the driver if device did no go through full hardware power cycle and POR bit is not marked. This was causing issues because if some cases driver should clean everything even if POR bit is not touched. Those cases are : During development or if device main power supply is 100% online but MCU reboot via RESET button or other and should get clear device.driver. For that reason always-on (AON) flag is introduced in MFD driver, as parent driver. RTc access it and act accordingly.

bogdanovs committed:
drivers: counter: rv3032c7: Add rv3032 counter driver

Expose timer functionality from rv-3032-c7 via counter api.
I2C communication and IRQs are arbitrated via MFD parent driver
for rtc, sensor and counter at once.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
drivers: sensor: rv3032c7: Add temperature sensor driver for rv3032

Expose internal temperature sensor for rv-3032-c7. Driver is reporting
temperature from internal deice sensors and have HIGH an LOW temperature
treshhold which are triggering interrupts. I2C communication and IRQ are
handled from parent MFD driver.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
drivers: mfd: rv3032: Add MFD driver for rv3032

Add MFD driver for managing rv3032 functionality which did not fit RTC
api. That way are implemented part which otherwise will not be used:
- RTC
- counter/timer
- temperature sensor

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
drivers: rtc: Kconfig: Update rv3032 Kconfig to utilize MFD driver

Add dependecy for MFD driver and remove I2C dependency since it will
be managed from MFD driver together with interrupts.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
dts: bindings: rtc: Update rv3032 bindings

Remove gpio-int since interrupts will be managed from MFD driver
insted, acticting like dispatcher for interrupts.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
dts: bindings: rv3032: Remove backup from rtc driver

Remove backup flag from RTC driver and move it to parent instead.
RTC driver still can access the values of the flags but MFD need
to know what is expected in term of initial behaviour.
Some assumptions done in RTC driver are causing inconsistant behaviour.
We cannot assume that the RTC alarms and notifications need to be
cleaned only if POR bit is triggered. We can have situation where
whole system go throigh soft restart and need to zero RV3032, so driver
need to have option to allow that kind of behaviour.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
drivers: rtc: rv3032: Rework driver to use MFD driver

Rework rv3032 implementaion to use MFD driver unstead of local
implementaion for I2C access and Interrupt managements. This way
could be utilized parts which are not implemented because did not fit
in to RTC api - counter and temperature sensor. To have propper
sincronization arbitration is need which is handled from mfd_rv3032.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
tests: rtc: Update microcrystal,rv3032 in build_all test suite

Update rv3032 to reflect latest chanes to the driver.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
tests: counter: Add microcrystal,rv3032-counter in build_all test suite

Ensure this driver is tested in the build_all test suite.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
tests: sensor: Add microcrystal,rv3032-temp in build_all test suite

Ensure this driver is tested in the build_all test suite.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bjarki-andreasen commented on drivers/counter/Kconfig.rv3032_counter:
```diff
@@ -0,0 +1,17 @@
+	bool "RV3032 Counter driver"
+	default y
+	depends on DT_HAS_MICROCRYSTAL_RV3032_COUNTER_ENABLED
+	depends on MFD_RV3032
```
This should be a select

bjarki-andreasen replied:
That's what I want to achieve with the select, the RTC, sensor, or counter, require the parent to be selected, so if you try to enable only one of the child devices, but the parent is not enabled (status="disabled" or MFD=n), that is a bug, its a direct dependency.

bjarki-andreasen commented on drivers/mfd/Kconfig.rv3032_mfd:
```diff
@@ -0,0 +1,18 @@
+	bool "MFD driver for RV3032"
+	default y
+	depends on DT_HAS_MICROCRYSTAL_RV3032_MFD_ENABLED
+	depends on I2C
```
select I2C

bjarki-andreasen reviewed (changes requested):
Some nits, looks really good!

bogdanovs commented:
> Some nits, looks really good!

Thanks for review , all solved.
Rebase on latest main and drop last commit which was solving CI issue, now is merged in main.

bjarki-andreasen reviewed (dismissed):
Cool

bogdanovs committed:
tests: build_all: counter: Enable GPIO in prj.conf

Add missing CONFIG_GPIO in prj.conf which cause fails
while testing drivers which use GPIOs

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

MaureenHelm commented on dts/bindings/sensor/microcrystal,rv3032-temp.yaml:
```diff
@@ -0,0 +1,20 @@
+properties:
+  low-temp:
+    type: int
+    default: -40
```
Need to explain in the description why the default value was selected
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

MaureenHelm reviewed (changes requested):
One minor issue, but otherwise LGTM

bogdanovs commented on dts/bindings/sensor/microcrystal,rv3032-temp.yaml:
```diff
@@ -0,0 +1,20 @@
+      Low temperature value
+  high-temp:
+    type: int
+    default: 85
```
Thanks, done!
